### PR TITLE
[WB-9774] Launch agent support instance roles for ec2 and eks

### PR DESCRIPTION
**Pull request by KyleGoyette**
[Fixes WB-9774](https://wandb.atlassian.net/browse/WB-9774)
Fixes #NNNN

Description
-----------
Tightens up how the launch agent works with instance roles for AWS in both EC2 and on EKS. The associated role permissions are shown in this doc: https://www.notion.so/wandbai/Launch-On-AWS-d0ae59a93e04445c9c586c5436c695d7

Testing
-------
On EC2, EKS and locally to make sure nothing broke. Lots of manual testing.

Checklist
-------
- Name PR "[WB-NNNN][WB-MMMM] Add support for..." similar to entries in CHANGELOG.md
- Include reference to internal ticket "Fixes WB-NNNN" (and github issue "Fixes #NNNN" if applicable)

### CODE REVIEW

File: wandb/sdk/launch/deploys/aws-eks/launch-config.yaml [hunk 1]
```diff
@@ -74,16 +74,21 @@ roleRef:
 apiVersion: v1
 data:
   launch-config.yaml: |
-    base_url: https://api.wandb.ai # TODO: SET YOUR WANDB BASE URL
-    project: launch-welcome # TODO: SET YOUR ML PROJECT
-    max_jobs: -1 # TODO: SET YOUR MAX CONCURRENT JOBS HERE
+    base_url: https://api.wandb.ai # TODO: SET WANDB BASE URL
```
Comment (hu-po) [inline wandb/sdk/launch/deploys/aws-eks/launch-config.yaml:77]: The allcaps kinda scares me here ngl

File: wandb/sdk/launch/launch.py [hunk 2]
```diff
@@ -37,6 +37,7 @@ def resolve_agent_config(
         "base_url": api.settings("base_url"),
         "registry": {},
         "build": {},
+        "runner": {},
```
Comment (hu-po) [inline wandb/sdk/launch/launch.py:40]: default `type` field?

File: wandb/sdk/launch/utils.py [hunk 3]
```diff
@@ -198,7 +198,7 @@ def is_bare_wandb_uri(uri: str) -> bool:
     result = uri.split("/")[1:]
```
Comment (hu-po) [inline wandb/sdk/launch/utils.py:198]: If there is no `/` in the uri this will error out. Maybe want to catch that and return `False` in that case.

File: wandb/sdk/launch/deploys/aws-eks/launch-agent.yaml [hunk 4]
```diff
@@ -16,7 +16,7 @@ spec:
       serviceAccountName: wandb-launch-serviceaccount
       containers:
       - name: launch-agent
-        image: wandb/launch-agent-dev:kaniko-test6
+        image: wandb/launch-agent-dev:latest
```
[the remark this answers is not in the document]
Comment (KyleGoyette) [reply wandb/sdk/launch/deploys/aws-eks/launch-agent.yaml:19]: No, let's keep it automatically updating

File: wandb/sdk/launch/deploys/aws-eks/launch-config.yaml (hunk 1 above)
Comment (KyleGoyette) [reply wandb/sdk/launch/deploys/aws-eks/launch-config.yaml:77]: Fixed

File: wandb/sdk/launch/utils.py (hunk 3 above)
Comment (KyleGoyette) [reply wandb/sdk/launch/utils.py:202]: There is already a check on the line before that ensures there's at least one "/"

File: wandb/sdk/launch/builder/kaniko.py [hunk 5]
```diff
@@ -98,7 +99,7 @@ def __init__(self, builder_config: Dict[str, Any]):
     def _create_docker_ecr_config_map(
         self, corev1_client: client.CoreV1Api, repository: str
     ) -> None:
-        if self.cloud_provider == "AWS":
+        if self.cloud_provider == "aws":
```
[the remark this answers is not in the document]
Comment (KyleGoyette) [reply wandb/sdk/launch/builder/kaniko.py:102]: Nice catch

File: tests/tests_launch/test_launch_aws.py [hunk 6]
```diff
@@ -32,39 +35,108 @@ def mock_create_training_job(*args, **kwargs):
...
-        return ecr_client
+
+        return mock_ecr_client()
     elif args[0] == "sts":
         sts_client = MagicMock()
-        sts_client.get_caller_identity.return_value = {
-            "Account": "123456789012",
-        }
+        sts_client.get_caller_identity.return_value = {"Account": "123456789012"}
+
```
[the remark this answers is not in the document]
Comment (KyleGoyette) [reply tests/tests_launch/test_launch_aws.py]: nice

File: tests/unit_tests/tests_launch/test_launch_aws.py [hunk 7]
```diff
@@ -32,37 +35,105 @@ def mock_create_training_job(*args, **kwargs):
...
+    if kwargs.get("aws_access_key_id") is None:
         sts_client = MagicMock()
-        sts_client.get_caller_identity.return_value = {
-            "Account": "123456789012",
-        }
-        return sts_client
+        sts_client.get_caller_identity = MagicMock(
+            side_effect=botocore.exceptions.NoCredentialsError,
+        )
+    elif kwargs.get("aws_access_key_id") is not None:
```
Comment (hu-po) [inline tests/unit_tests/tests_launch/test_launch_aws.py:87]: could just be `else` but this makes it explicit so probably better for readability?

File: wandb/sdk/launch/builder/kaniko.py [hunk 8]
```diff
@@ -140,8 +141,7 @@ def _upload_build_context(self, run_id: str, context_path: str) -> str:
         with tarfile.TarFile.open(fileobj=context_file, mode="w:gz") as context_tgz:
             context_tgz.add(context_path, arcname=".")
         context_file.close()
-
-        if self.builder_config.get("cloud-provider") == "AWS":
+        if self.cloud_provider == "aws":
```
Comment (hu-po) [inline wandb/sdk/launch/builder/kaniko.py:144]: `.lower()`

File: wandb/sdk/launch/builder/kaniko.py [hunk 9]
```diff
@@ -163,7 +163,7 @@ def _upload_build_context(self, run_id: str, context_path: str) -> str:
                 raise LaunchError(f"Failed to upload build context to S3: {e}")
             return f"s3://{self.build_context_store}/{run_id}.tgz"
         # TODO: support gcp and azure cloud providers
-        elif self.builder_config.get("cloud-provider") == "gcp":
+        elif self.cloud_provider == "gcp":
```
Comment (hu-po) [inline wandb/sdk/launch/builder/kaniko.py:166]: also `.lower()`

File: wandb/sdk/launch/builder/kaniko.py [hunk 10]
```diff
@@ -267,7 +267,7 @@ def _create_kaniko_job(
         build_context_path: str,
     ) -> "client.V1Job":
         env = None
-        if self.instance_mode and self.cloud_provider == "AWS":
+        if self.instance_mode and self.cloud_provider == "aws":
```
Comment (hu-po) [inline wandb/sdk/launch/builder/kaniko.py:270]: also `.lower()`

File: wandb/sdk/launch/runner/aws.py [hunk 11]
```diff
@@ -200,13 +210,15 @@ def run(
...
 
         _logger.info("Connecting to sagemaker client")
-
-        sagemaker_client = boto3.client(
-            "sagemaker",
-            region_name=region,
-            aws_access_key_id=access_key,
-            aws_secret_access_key=secret_key,
-        )
+        if instance_role:
```
Comment (hu-po) [inline wandb/sdk/launch/runner/aws.py:213]: is not None ?

**Review by hu-po — APPROVED** [review]
Lots of intricate config/dict parsing, but I think catching all these edge cases is good.

File: tests/unit_tests/tests_launch/test_launch_aws.py (hunk 7 above)
Comment (KyleGoyette) [reply tests/unit_tests/tests_launch/test_launch_aws.py:87]: that was my take.

File: tests/unit_tests/tests_launch/test_launch_aws.py [hunk 12]
```diff
@@ -32,37 +35,105 @@ def mock_create_training_job(*args, **kwargs):
...
+        sts_client = MagicMock()
+        sts_client.get_caller_identity.return_value = {"Account": "123456789012"}
+
+    clients = {
+        "sagemaker": mock_sagemaker_client(),
+        "ecr": mock_ecr_client(),
+        "sts": sts_client,
+    }
+    return clients[args[0]]
```
[the remark this answers is not in the document]
Comment (KyleGoyette) [reply tests/unit_tests/tests_launch/test_launch_aws.py:96]: sounds good

File: wandb/sdk/launch/runner/aws.py (hunk 11 above)
Comment (KyleGoyette) [reply wandb/sdk/launch/runner/aws.py:213]: instance_role is a boolean.
